hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java

```diff
@@ -627,6 +627,11 @@ public static void verifyKeyNameWithSnapshotReservedWord(String keyName)
           "Cannot create key under path reserved for "
               + "snapshot: " + OM_SNAPSHOT_INDICATOR + OM_KEY_PREFIX,
               OMException.ResultCodes.INVALID_KEY_NAME);
+    } else if (keyName != null &&
+        keyName.equals(OM_SNAPSHOT_INDICATOR)) {
```

Comment:
In almost every case, this check is not going to throw, as the key will be valid. This change adds a second check, so we first check:

startsWith(".snapshot/")

Then we have the new check:

key.equals(".snapshot")

I think it would be more efficient to have something like:

```
  public static void verifyKeyNameWithSnapshotReservedWord(String keyName)
          throws OMException {

    if (keyName != null &&
        keyName.startsWith(OM_SNAPSHOT_INDICATOR)) {
      if (keyName.length() > OM_SNAPSHOT_INDICATOR.length()) {
        if (keyName.substring(OM_SNAPSHOT_INDICATOR.length())
            .startsWith(OM_KEY_PREFIX)) {
          throw new OMException(
              "Invalid key name, key name cannot start with "
                  + OM_SNAPSHOT_INDICATOR + " and followed by "
                  + "characters other than " + OM_KEY_PREFIX,
              OMException.ResultCodes.INVALID_KEY_NAME);
        }
      } else {
        // We checked for startsWith OM_SNAPSHOT_INDICATOR and the length is
        // the same, so it must be equal OM_SNAPSHOT_INDICATOR.
        throw new OMException(
            "Cannot create key with reserved name: " + OM_SNAPSHOT_INDICATOR,
            OMException.ResultCodes.INVALID_KEY_NAME);
      }
    }
  }
```

That way, for the vast majority of cases we check only that the key starts with ".snapshot" once and then do extra validation if it does start with .snapshot, which should be rare.

Reply:
+1 on @sodonnel

Reply:
Changed the logic as suggested